sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -137,45 +137,50 @@ case class CaseWhen(branches: Seq[(Expression, Expression)], elseValue: Option[E
   }
 
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
-    val got = ctx.freshName("got")
+    val len = branches.length
 
-    val cases = branches.map { case (condition, value) =>
-      val cond = condition.gen(ctx)
-      val res = value.gen(ctx)
-      s"""
-        if (!$got) {
+    def genBranch(branchIndex: Int): String =
```

Comment:
this code is pretty confusing. do you think you can simplify the flow a bit?

maybe it'd help to not do this with recursion, and also have the else case outside (if you want to have it inside, it should be at the end rather than in the beginning because else is conceptually the last case, not the first)